contracts/utils/structs/MerkleTree.sol

```diff
@@ -0,0 +1,160 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import {Hashes} from "../cryptography/Hashes.sol";
+import {Arrays} from "../Arrays.sol";
+import {Panic} from "../Panic.sol";
+
+/**
+ * @dev A complete binary tree with the ability to sequentially insert leaves, changing them from a zero to a non-zero
+ * value, while keeping a history of merkle roots. This structure allows inserting commitment (or other entries) that
+ * are not stored, but can be proven to be part of the tree.
+ *
+ * The history of merkle roots allow inclusion proofs to remain valid even if leaves are inserted into the tree between
+ * the moment the proof is generated and the moment it's verified.
+ *
+ * Each tree can be customized to use specific
+ * - depth
+ * - length of the root history
+ * - zero values (for "empty" leaves)
+ * - hash function
+ *
+ * IMPORTANT: By design, the tree include zero leaves. Customizing the "zero value" might be necessary to ensure that
+ * empty leaves being provably part of the tree is not a security issue.
+ *
+ * _Available since v5.1._
+ */
+library MerkleTree {
+    /**
+     * @dev Maximum supported depth. Beyond that, some checks will fail to properly work.
+     * This should be enough for any realistic usecase.
+     */
+    uint256 private constant MAX_DEPTH = 255;
+
+    /**
+     * @dev Merkle tree cannot be set-up because requested depth is to large.
+     */
+    error MerkleTreeInvalidDepth(uint256 depth, uint256 maxDepth);
+
+    /**
+     * @dev The `sides` and `zero` arrays are set, at initialization, to have a length equal to the depth of the tree.
+     * No push/pop operations should be performed of these arrays, and their lengths should not be updated.
+     *
+     * The hashing function used during initialization to compute the `zeros` values (value of a node at a given depth
+     * for which the subtree is full of zero leaves). This function is kept in the structure for handling insertions.
+     *
+     * Contracts using this structure may want to use a secondary structure to store a (partial) list of historical
+     * roots. This could be done using a circular buffer (to keep the last N roots) or the {Checkpoints} library to
+     * keep a more complete history. Note that if using the Checkpoints.Trace224 structure for storing roots, you will
+     * be limited to keeping "only" 26 bytes out of the root's 32. This should not be a security issue.
+     */
+    struct Bytes32MerkleTree {
+        bytes32 root;
+        uint256 nextLeafIndex;
+        bytes32[] sides;
+        bytes32[] zeros;
```

Comment:
> I think it would be better if there was a way to precompute the zeros in constants or immutable variables so each access is cheaper.

immutable onyl support value type. Here we are dealing with an array which length is only known at runtime (after compilation). Immutable (and constant) don't support that.

Reply:
> I think the "zero" value is only used to avoid the actual zero being a provable part of the tree.

I've seen other zero used. In particular ones that are produced by doing `hash("")` with hash being any hashing function (keccak256, MiMc, Pedersen, ...)

Reply:
> Immutable (and constant) don't support that.

Yeah, and the possible combinations are too much to procedurally generate them. I think it's not worth pursuing this, but hey it was almost a good idea lol

> I've seen other zero used. In particular ones that are produced by doing hash("")

Would you mind elaborating on the use cases of this? Sounds interesting